### PR TITLE
Refactor participant pipeline to components

### PR DESCRIPTION
**Pull request by kaitj**
Chatted a little bit offline about this, but I was going in circles looking at the code trying to figure out where to start. I started to break down the `participant_raw` pipeline into components. This will (hopefully) make the pipelines a little easier to follow and refactor. Everything is up for discussion and to make changes as necessary!

Summary of the main changes below:
* Grouped the CLI arguments by analysis-level, which may provide more clarity in the usage / help message.
* Also renamed `mriqc-dir` argument to just `qc-dir` to reflect that it can ingest any QC output in the "right" format.
* Refactored sub-module organization (noticable in `__main__`) for (I think) improved imports.  
* Moved the metric generation code to its own module under `niftyone`, making it an import.
* Organized figure generation into separate sub-modules by modality (e.g. `anat`, `func`, etc.). 
* Added (optional) unit tests for the new modules while aiming to minimize complex mocking. These tests are skipped on Github Actions and ignored by Codecov due to needing input data which is not pulled in in the actions environment.
* Removed suffix filter, allowing for broader QC options with warning messaging for unsupported suffixes.
* Switched `if-elif-else` calls to `match-case` where appropriate.

A few things to discuss:
* Naming of the figure generating module
* Figure generation grouping (e.g. some common elements across modalities or even within modlities)

I think the group pipeline can undergo something similar, but it probably won't be as big of a change.

### CODE REVIEW

**Comment by kaitj** [comment]
Good for a look - I'm still going to make 1 more push to rename `participant_raw` to `participant`.

File: pyproject.toml
Comment (clane9) [inline pyproject.toml:8]: What was the cause of these nonsignificant diffs? Some version update of ruff or something?

File: src/niftyone/__main__.py
Comment (clane9) [inline src/niftyone/__main__.py:69]: Adding the arg groups is nice 👍 .

File: src/niftyone/figures/anat.py
Comment (clane9) [inline src/niftyone/figures/anat.py:15]: Leaving a note from our offline conversation. Currently two figures plus qc metrics are being generated here. We discussed decoupling this a bit more. Specifically, we discussed:

1. Separating the concerns of figure generation and QC metric extraction.
2. Uncoupling the figures so that there is one generator per figure
3. Assigning the responsibility of filtering the records to the generator itself. (Currently that responsibility lies outside in the calling pipeline.)

One possible abstraction we discussed was:

```python
class Generator:
    def __init__(self, filters: dict[str, Any]):
        self.filters = filters

    def __call__(self, table: BIDSTable, out_dir: Path, overwrite: bool = False):
        """Filter table and generate figures for each matching record."""
        table = table.filter_multi(**self.filters)
        for _, record in table.iterrows():
            self.generate(record, out_dir, overwrite)

    def generate(self, record: pd.Series, out_dir: Path, overwrite: bool = False):
        """Generate figure for record."""
        raise NotImplementedError


class Pipeline:
    def __init__(self, generators: list[Generator]):
        self.generators = generators

    def run(self, table: BIDSTable, out_dir: Path, overwrite: bool = False):
        for gen in self.generators:
            gen(table, out_dir, overwrite)
```

For configuring pipelines, we could define the list of figure generators (and their kwargs) in a yaml config. Then at runtime, construct the list of generators and run the pipeline.

```yaml
figures:
  three_view:
    filters:
      suffix: T1w
  slice_video:
    filters:
      suffix: T1w
  three_view_video:
    filters:
      suffix: bold
  carpet_plot:
    filters:
      suffix: bold
  mean_std:
    filters:
      suffix: bold
```

The generator lookup could be handled with a global registry, with generators registered via a decorator when the class is defined.

I'm not fully convinced this is the best abstraction, so I'm open to alternatives. In particular, the config looks quite repetitive. This is not necessarily the worst thing, since decoupling is worth some repetition. But it might be a sign we are missing a better idea. 

Maybe something along the lines of letting groups of figures be associated with each set of BIDS `filters`, rather than a 1-1 mapping of filters to figures.

@nx10 thoughts?

**Review by clane9 — APPROVED** [review]
Thanks @kaitj! Happy to merge this now although it is more like part 1 of a longer API update.

I left one longer comment summarizing what I remember from our offline conversation. Lmk your thoughts.

File: pyproject.toml
Comment (kaitj) [reply pyproject.toml:8]: Must have been - I honestly didn't even notice this so it must have been formatted by the pre-commit.

**Comment by kaitj** [comment]
I actually ended up making a few more changes this morning (mostly minor), but to clean up the code further. I can wait until this is merged in to start with a clean diff (part 2 of longer API update? 😉)

**Comment by clane9** [comment]
If it fits with this PR it might as well go in here. You could convert this back to draft and then mark as ready once you're done if you want.

**Comment by kaitj** [comment]
Sure, I mean it is pretty much done so I can request a new review once pushed. 

One thing that I've been going back and forth on are the pipelines and the `launch` calls. Right now the `participant` and `group` level calls are under `pipelines`, while `launch` is under `niftyone` itself. Wondering if it makes sense just to put them all into a module like `analysis_level` or something of the sort. And have pipelines be reserved for part of the implementation of the above discussion.

**Comment by clane9** [comment]
Hm good point. I'm not totally sure, but my instinct is that we should not overdo the levels of abstraction. I think we really should only ever need two "pipelines", the static group pipeline and the configurable participant pipeline. Perhaps we keep the current organization, and get rid of the `Pipeline` abstraction I proposed above (instead pushing that logic into the singular participant pipeline), rather than introduce another "analysis_level" layer of abstraction on top of "pipeline".

You are right that launch being its own thing is probably not good organization. Perhaps it could be a third "pipeline"?

It would also be fine call participant/group/launch all "analysis_levels" rather than "pipelines". Getting the abstractions right is more important to me than the naming.

File: src/niftyone/figures/anat.py
Comment (kaitj) [reply src/niftyone/figures/anat.py:15]: Hmm, another (not fully flushed out) thought if going the yaml route:

```yaml
key1:
  filters:
      desc: [True, brain, False]
      suffix: [T1w, T2w, T1map]
  views: [three_view, slice_video]

key2:
  filters:
      suffix: bold
  views: [three_view_video, carpet_plot, mean_std]
```

The entities provided in filters are zipped and passed along together. If an entity `True`, it treats that entity as a wildcard (e.g), if `False`, exclude rows that contain an entity for that filter. Otherwise, the filter has to match exactly. If an entity isn't provided, we can also treat it similar to `False`. So in the above case, it would generate the views for `sub-<subject>_desc-*_T1w`, `sub-<subject>_desc-brain_T2w`, and `sub-<subject>_T1map`.

The views will indicate the figure(s) that should be generated for those matching filters. Figure generation time aside, the end-user can always ignore those in NiftyOne itself. The primary keys (e.g. `key1`, `key2`) are kind of arbitrary and really are only there to group filters with figures to generate. 

As an aside, we're going to need some good documentation so that a user can find the different views that can be generated (irrespective of the implementation method) 😅 .

And also open to other ideas of course.

File: pyproject.toml
Comment (nx10) [reply pyproject.toml]: Yeah this is the TOML formatter in our pre-commit

File: src/niftyone/figures/anat.py
Comment (nx10) [reply src/niftyone/figures/anat.py:15]: Overall I would prefer a [Pandas query string](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html) or [Polars SQL query](https://docs.pola.rs/py-polars/html/reference/api/polars.SQLContext.execute.html) (for SQL I would just let the user set `SELECT * FROM df WHERE {userinput}`). 

It's always tempting to reimplement a small query DSL because one thinks they don't need all of that complexity, but then slowly realize that the complexity is needed.

Comment (nx10) [reply src/niftyone/figures/anat.py:15]: With @kaitj s matrix groups this would look like:

```yaml
my_t1w_images:
  query: "( desc in (None, 'brain') ) and (suffix in ('T1w', 'T2w', 'T1map') )"
  views: [three_view, slice_video]

my_bold_images:
  query: "suffix == 'bold'"
  views: [three_view_video, carpet_plot, mean_std]
```

Group name could also become a tag, then the same logic can be used to generate user defined tags (potentially without views):

```yaml
really_bad_images:
  query: "whatever_qc_metric < 0"
```

Comment (clane9) [reply src/niftyone/figures/anat.py:15]: Thanks @kaitj, @nx10 for the good suggestions. Let's go with your guys' preferred direction. I also think having a `query` option is a good idea. Having a simple way to do basic filters might also be a good idea, but it's less clear to me what the api for this should be. So I think let's just implement the `query` option for now and decide later on if there are common verbose query patterns we would like to streamline.

The suggestion for QC metric queries is also interesting. On the one hand, I can imagine a use case for generating figures conditional on some QC value. E.g. certain expensive to generate detailed figures might only be useful for bad or borderline data. On the other, it could be a bit of an anti-pattern since filtering by QC values can happen naturally within the fiftyone app. Also, it would require joining the QC metric table(s) with the b2t table prior to figure generation. (Not a big deal but something to note.) @kaitj what do you think about QC based queries?

Comment (kaitj) [reply src/niftyone/figures/anat.py:15]: Yeah, I like the `query` option too. I think it'll be familiar too (hopefully) for what I imagine would be the majority of users.

Regarding the QC metric queries, I think we are on a similar page. I can see a use case for it too. In fact, it is likely what we would want to do for identifying problematic images as you noted, but I also think it is probably better to leave that kind of filteirng for within the app or potentially for power(?) users (e.g. [interactive plot](https://docs.voxel51.com/user_guide/plots.html) sort of thing).

**Pull request merged**
